Class to read OpenDocument Tables

detrout opened the pull request:
closes #2311 

This is primarily intended for LibreOffice calc spreadsheets but will
also work with LO Writer and probably with LO Impress documents.

This is an alternate solution to https://github.com/pandas-dev/pandas/pull/9070
There are test cases with several different problematic LibreOffice spread sheets.
git diff upstream/master -u | flake8 appeared to pass.

... I didn't do the whats new entry. Though I expect there's some more work to do before submitting this. I just wanted to get the core code in for comments.

The open issues is, the workaround for https://github.com/pandas-dev/pandas/issues/25422 is embedded in the current code (so all my tests pass right now) but that, or a better solution should move closer to the iso 8601 parser.

Also I don't have the parser class hooked up to a read_excel or read_ods function.

Using read_excel bothers me some... because its not an excel file. I think it would be more clear if there was either a separate read_ods function or a generic read_spreadsheet function than just pressing read_excel into being a generic spreadsheet reader.

Also this just a reader. I had never gotten around to implementing a writer.

pep8speaks commented:
Hello @detrout! Thanks for submitting the PR.





 - In the file [`pandas/tests/io/test_opendocument.py`](https://github.com/pandas-dev/pandas/blob/7e6a59ccf08f573adbb3bc6befde98791e86b67a/pandas/tests/io/test_opendocument.py), following are the PEP8 issues :

> [Line 147:33](https://github.com/pandas-dev/pandas/blob/7e6a59ccf08f573adbb3bc6befde98791e86b67a/pandas/tests/io/test_opendocument.py#L147): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 148:33](https://github.com/pandas-dev/pandas/blob/7e6a59ccf08f573adbb3bc6befde98791e86b67a/pandas/tests/io/test_opendocument.py#L148): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 149:33](https://github.com/pandas-dev/pandas/blob/7e6a59ccf08f573adbb3bc6befde98791e86b67a/pandas/tests/io/test_opendocument.py#L149): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator

detrout commented:
Ah the coverage went down but that's because the tests couldn't run because odfpy wasn't available.

Also I fixed the whitespace problem.

jreback reviewed (changes requested):
so you are going to want to put this in pandas/io/excel and model it after one of the readers maybe _xlrd.py

tests in s similar way

then it will just work; read_excel is the generic interface here (we can discuss in a new issue whether aliasing to read_spreadsheet is ok)

detrout commented:
Wait a second... I just realized that read_excel already supports using two totally different libraries for reading .xls and .xlsx files. Now it makes more sense why you want it hooked into io.excel.

OK it'll take me a bit to understand how the hook works.

My current reader class is ODFReader, but ODFFile or OpenDocumentFile might be more consistent with ExcelFile. Do you have any preferences?

Lastly is there any place I can add then odfpy dependency so the ci framework will be able to run my tests?

Thanks you for your comments.
Diane

jreback commented:
> My current reader class is ODFReader, but ODFFile or OpenDocumentFile might be more consistent with ExcelFile. Do you have any preferences?

you don't need to expose any of this directly. These can all be internal. invoking ``pd.read_excel(...., engine='ods')`` should just work.

jreback commented:
deps you would add in the ci/deps/* files; don't add to *every* one as you also have to function w/o being included (the framework handles this for you though)

detrout committed:
Class to read OpenDocument Tables

This is primarly intended for OpenDocument spreadsheets like what is
generated by LibreOffice calc but will also work with LibreOffice
Writer.

detrout commented:
Ok everything is moved into pandas.io.excel and tests/io/test_excel.py

Currently its triggered by passing engine='odf' to read_excel. It looks like pandas is currently depending on xlrd to figure out what to do with the various excel style file extensions.

Should there be something to activate odf reading in ExcelFile.__init__ based on file extension or trying xlrd parsing and catching the error?

detrout commented:
Drat. I forgot the CI dependency change.

detrout commented:
Ok I added odfpy to the travis-36.yaml file since that seemed to be a thorough test configuration.

jreback commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+    filepath_or_stream: string, path to be parsed or
+        an open readable stream.
+    """
+    def __init__(self, filepath_or_stream):
```
call this filepath_or_buffer

jreback commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+    def get_sheet(self, name):
+        """Given a sheet name or index, return the root ODF Table node
+        """
+        if isinstance(name, str):
```
use compat.string_types

jreback commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+import pandas
```
import pandas as pd

jreback commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+        parser = TextParser(data, **kwds)
+        return parser.read()
+
+    def __get_table(self, sheet):
```
use a single underscore for private methods

jreback reviewed (changes requested):
@detrout looks really good! pls add a whatsnew note, New features in 0.25.0, also pls update the io.rst documentation as needed

jreback commented:
references the issue #2311 in your whatsnew

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+            raise ImportError("Install odfpy for OpenDocument support")
+
+        self.filepath_or_stream = None
+        self.document = None
```
Do we need this here since a subsequent assignment happens a few lines down?

detrout replied:
Nope, the two assignments were a left over from my standalone implementation where loading happened in a separate function.

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+                'Unrecognized sheet identifier type {}. Please use'
+                'a string or integer'.format(type(name)))
+
+    def parse(self, sheet_name=0, **kwds):
```
This is already defined in the base class so ideally don't need to override here

detrout replied:
Ok. _ODFReader wasn't derived from _BaseExcelReader... and I have a problem making it fit. The sheet object returned by xlrd has properties that I can't easily match.

As far as I can tell from looking at the content.xml the only way to know the nrows, which is currently used by _BaseExcelReader is to actually parse the odf table.

WillAyd replied:
Is `nrows` the only hang up here? Had a similar discussion here:

https://github.com/pandas-dev/pandas/pull/25092/files#r255718071

So if it's a hang up for two open PRs might consider doing that as a precursor to simplify things all around

detrout replied:
That's a good question. I'm pretty tired right now I'll try looking if something other than nrows might go also break in a few days.

The convert_float parameter for get_sheet_data doesn't make as much sense as ODF has separate types for integer and floats. (Here's the list of supported cell types: http://docs.oasis-open.org/office/v1.2/os/OpenDocument-v1.2-os-part1.html#__RefHeading__1417680_253892949 )

jreback replied:
so its ok to ignore a passed parameter; if its in conflict then you could raise an error if its explicity passed

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+        i = self.sheet_names.index(name)
+        return self.__get_table(self.tables[i])
+
+    def get_sheet(self, name):
```
Do we even need this method? I see it's added as a result of the subclassed `parse` method but the behavior here to get a sheet via inference of position / name is not something I think the other engine has, so don't want to introduce anything inconsistent

detrout replied:
Pending figuring out how to relate _ODFReader to the current _BaseExcelFile and if there's a way to make the base class parse function work, I renamed this to _get_sheet to indicate its private.

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+            from odf.opendocument import load as document_load
+            from odf.table import Table
+        except ImportError:
+            raise ImportError("Install odfpy for OpenDocument support")
```
Is there a minimum version required to match the structure of the imports being performed in this code?

detrout replied:
I'm not sure the exact version but as far as I can tell I started with a 1.3.x version and the API looks stable since then. Also pypi suggests that 1.2.0 was a "pre-release" https://pypi.org/project/odfpy/#history

I did add the text to match what's in _xlrd. so I'm going to mark this issue was resolved.

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,165 @@
+        self.document = None
+        self.tables = None
+        self.filepath_or_stream = filepath_or_stream
+        self.document = document_load(filepath_or_stream)
```
This is called `self.book` in the existing xlrd code - can we use the same naming convention here?

detrout replied:
That's a little harder case... in _xlrd._XlrdReader book looks to be a reference to the xlrd class that provides the interface to excel files.

odfpy provides more of an ElementTree or lxml level interface than anything useful about tables. self.tables is just a dictionary pointing to the root node of table XML nodes found in the OpenDocument file. I'm not sure its high level enough to be considered equivalent to a "book" object.

WillAyd replied:
OK - you'll have to excuse that I'm not terribly familiar with odfpy. This is a rather minor sticking point so not really a blocker for this PR but consistency in the codebase is very important; it *may* make sense subsequently to change the xlrd code to reflect this convention or come up with an abstraction for both that reads consistently

WillAyd commented on pandas/tests/io/test_excel.py:
```diff
@@ -2557,3 +2557,135 @@ def test_excelwriter_fspath(self):
+        pth = os.path.join(self.dirpath, 'datatypes.ods')
+        book = _ODFReader(pth)
+
+        with pytest.raises(ValueError):
```
Can you move methods that should raise into a separate test called `test_get_sheet_raises`?

WillAyd commented on pandas/tests/io/test_excel.py:
```diff
@@ -2557,3 +2557,135 @@ def test_excelwriter_fspath(self):
+        assert book.sheet_names == ['Sheet1']
+
+    def test_read_types(self):
+        """Make sure we read ODF data types correctly
```
OK to remove docstring - I don't think we have these for any other tests

detrout replied:
Some of the docstrings were just restating the function name, but some of them are descriptions of what the test is about. I'd like those more detailed comments to stay, is it ok if they're docstrings or should I turn them into # comments?

Like:
```
    def test_read_invalid_types(self):
        """Make sure we throw an exception when encountering a new value-type

        I had to manually create an invalid ods file by directly
        editing the extracted xml. So it probably won't open in
        LibreOffice correctly.
        """
```

WillAyd replied:
Comments are fine though as minimal as possible is preferred

WillAyd commented on pandas/tests/io/test_excel.py:
```diff
@@ -2557,3 +2557,135 @@ def test_excelwriter_fspath(self):
+             ['UBERON:0002101']])
+        tm.assert_equal(sheet, expected)
+
+    def test_read_invalid_types(self):
```
Append `_raises` to test name

WillAyd commented on pandas/tests/io/test_excel.py:
```diff
@@ -2557,3 +2557,135 @@ def test_excelwriter_fspath(self):
+        tm.assert_equal(sheet, expected)
+
+    def test_read_invalid_types(self):
+        """Make sure we throw an exception when encountering a new value-type
```
Also no docstring

detrout replied:
I'd really like the discussion about the invalid file to stick around somehow. (It fails xml validation)

WillAyd commented on pandas/tests/io/test_excel.py:
```diff
@@ -2557,3 +2557,135 @@ def test_excelwriter_fspath(self):
             assert os.fspath(writer) == str(path)
+
+
+@td.skip_if_no('odf')
```
Ideally you should add `odf` as an item here:

https://github.com/pandas-dev/pandas/blob/fe1654faa86836a0007bb513504e57c5c9935b8b/pandas/tests/io/test_excel.py#L121

That way all base tests get executed. You might need a skip mark for when things aren't installed. Right now this is done only for xlrd via the class decorator of `SharedItems` but you may need to move that here as well to get this to work

detrout replied:
I don't understand what the fixture magic is doing here yet.

WillAyd replied:
Hmm fair enough. So the fixture is basically allowing all of the SharedItems tests to be run twice, once with an engine explicitly set as `xlrd` and the other with an engine of `None`. 

The inheritance here is kind of tricky because all of the SharedItems operate on actual excel files, which this isn't doing. Not sure the best way to consolidate so tabling this comment for now

WillAyd reviewed (changes requested):
Here are some comments from my side but agreed this looks very nice - kudos!

detrout committed:
Remove unneeded assignments

detrout committed:
Rename filepath_or_stream to filepath_or_buffer

detrout committed:
Use compat.string_types instead of str

detrout committed:
Use pd as name as pandas

detrout committed:
Use single underscore for private functions

detrout committed:
Return an unparsed sheet.

xlrd's versions of get_sheet_by_index return unparsed
sheet objects which the parse function then feeds to get_sheet_data to
actually turn into a DataFrame

Use get_sheet_data as name of function that converts a sheet
into a structure to be passed to TextParser. (To be more consistent
with xlrd)

detrout committed:
Move ODFReader get_sheet exception testing code to its own function

that ends with _raises

detrout committed:
Append _raises to end of function name that tests exceptions

detrout committed:
Remove test docstrings that include no useful information

detrout committed:
Indicate likely minimum version.

I started this a couple of years ago and so was using a 1.3ish
release. Looking through the repository, it looks like the APIs
I'm using are stable since then.

detrout commented:
Ok I resolved many of the issues and pushed them as separate commits for the moment so its easier to see what I did.

There's a few issues I'd like to discuss further, and I left those issues open.

jreback commented:
can you merge master

jreback commented:
@detrout can you merge master and update; let's see if we can get this in

detrout commented:
Sorry I got distracted. I'll try to do it tomorrow.

detrout committed:
Convert notes about some OpenDocument tests to comments

detrout committed:
Add note about new OpenDocument functionality to whatsnew

detrout commented:
Ok I rebased to master. Convered all my important test docstrings to comments and added a whatsnew for v0.25 about the OpenDocument support.

I hope I remembered to fix everything (and that rebasing was the right merge strategy)

Also since you use python as a macro language for LibreOffice, at some point I'd like to extend my ODFReader class to support reading tables while running in LO. (I only tried it once, and had problems with importing pandas in the LO environment, so it's going to be a while before that works. For now those experiments are going to stay in my [detrout/pandasodf](https://github.com/detrout/pandasodf) repository)

WillAyd commented:
@detrout looks like there are still some conflicts. Workflow to resolve locally on your branch should be as follows:

```sh
git fetch upstream
git merge upstream/master
# Resolve conflicts here...
git merge --continue
git push origin YOUR_BRANCH_NAME
```

detrout committed:
Sort imports correctly

detrout commented:
I'm not sure what to do about this travis failure?

```
self = <pandas.io.excel._odfreader._ODFReader object at 0x7fbefb3f1c18>
name = 0
    def _get_sheet(self, name):
        """Given a sheet name or index, return the root ODF Table node
        """
>       if isinstance(name, compat.string_types):
E       AttributeError: module 'pandas.compat' has no attribute 'string_types'
pandas/io/excel/_odfreader.py:42: AttributeError
```

You asked me to use pandas.compat.string_types, and its present in 0.23.3, but it looks gone in 0.25. There's just a couple of python calls to isinstance(source, bastring) or (isinstance(string, str)

jreback commented:
@detrout just use
``isinstance(name, str)`` as the string_types has been removed now that we only support PY3

detrout committed:
Use str instead of compat.string_types

Pandas dropped support for Python 2.

WillAyd commented:
@detrout can you merge master? We've had quite a few changes between dropping Py2 support and refactoring some of the Excel classes that come into play here.

This feature is certainly welcome so hoping we can get this pushed through!

detrout commented:
Ok ... I'll try to get to it in the next day or two. I might be back with some questions.

Or is there a recommended place to look for guidance on the desired changes?

WillAyd commented:
Feel free to ask here or on Gitter if you need help

detrout commented:
Ok thanks.

detrout committed:
Remove leading underscore from ODFParser

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,164 @@
+from pandas.io.parsers import TextParser
+
+
+class ODFReader(object):
```
Should remove `(object)` here - we recently get rid of the Python2 "new style" classes so keeping this is what is causing your CI failure

detrout replied:
Thanks for the hint...

It first complained that I had a leading underscore in _ODFReader which i fixed, but I guess there were more lint checks hiding behind that one.

detrout committed:
Remove obsolete class (object)

detrout commented:
As far as I could tell the test failures in continuous-integration/travis-ci/pr look like they're from deprecation warnings in xlrd?

WillAyd commented:
@detrout seems to be the case; not sure on first glance why they would start appearing here

detrout commented:
> @detrout seems to be the case; not sure on first glance why they would start appearing here

I don't know. I did find this though:
https://github.com/pandas-dev/pandas/issues/22682

Also in the travis log there's a variable PANDAS_TESTING_MODE="deprecation". Does that enable deprecation testing? Do the other test configurations disable checking deprecations?

WillAyd commented:
Hmm so I'm not able to reproduce locally with xlrd 1.2 and python 3.7.2; must be a combination of some factors at play here

> Also in the travis log there's a variable PANDAS_TESTING_MODE="deprecation". Does that enable deprecation testing? Do the other test configurations disable checking deprecations?

That's for a different build so not applicable in this case. Something weird going on we'll just have to figure out

WillAyd commented:
Can confirm locally that these warnings appear in Python 3.6.8:

<details>

INSTALLED VERSIONS
------------------
commit: 223f7caad35412afbedaa76b1ca40fdb10dbb0a2
python: 3.6.8.final.0
python-bits: 64
OS: Darwin
OS-release: 18.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.25.0.dev0+589.g223f7caad
pytest: 4.5.0
pip: 19.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.2.0
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

Interestingly enough downgrading to xlrd 1.1 fixes the issue:

<details>

INSTALLED VERSIONS
------------------
commit: 223f7caad35412afbedaa76b1ca40fdb10dbb0a2
python: 3.6.8.final.0
python-bits: 64
OS: Darwin
OS-release: 18.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.25.0.dev0+589.g223f7caad
pytest: 4.5.0
pip: 19.1
setuptools: 41.0.1
Cython: 0.29.7
numpy: 1.16.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

WillAyd commented:
Ah so I've figured it out. Ignore my previous comment as my environments were not properly configured.

The problem here is that `odfpy` installs defusedxml as a dependency. There is some code on the xlrd side which prioritizes imports and sets global state as to whether or not iteration is supported.

Here's the import machinery:

https://github.com/python-excel/xlrd/blob/c5a95aedce43a5997e76e468e286d85ef035b173/xlrd/xlsx.py#L39

Here's the global state setting, which fails with defusedXML and throws all the warnings subsequently:

https://github.com/python-excel/xlrd/blob/c5a95aedce43a5997e76e468e286d85ef035b173/xlrd/xlsx.py#L59

Not sure what the best course of action is just yet. This is pretty nuanced and would throw a lot of warnings on the user side of things if they install xlrd and defusedxml together

WillAyd commented:
So the issue on the xlrd was closed in #315 and per #26487 xlrd is unmaintained so there may not be a way around this. I think will need to filter those warnings in the tests and maybe add a whatsnew note about how installing odfpy (specifically defusedxml) alongside xlrd may generate those warnings

detrout commented:
Hm. I read https://groups.google.com/forum/#!msg/python-excel/P6TjJgFVjMI/g8d0eWxTBQAJ
and that seems to suggest if someone were to propose a fix, it might get merged.

I wonder how hard it'd be to resolve the compatibility issue between xlrd and defusedxml.

WillAyd commented:
There was a proposal on the xlrd board but they just closed out all issues and suggested we move away from the package so I don’t think there will be a good level of support going forward. 

Even if fixed I don’t think we would want to immediately bump version support so something well have to live with

jreback commented:
xlrd has been stable for quite a long time
so i don’t think anything urgent

WillAyd commented:
@detrout can you merge master? The excel tests were just overhauled so should be easier to integrate now. May just need to catch the warning with xml deprecations

detrout committed:
Improve docstring text

detrout commented:
I'd been more busy than usual with work, but I just pushed my rebased against master. (And fixed a badly written docstring)

detrout commented:
Hm. Need to double check those test cases though...

WillAyd commented on pandas/tests/io/test_excel.py:
```diff
@@ -2556,3 +2556,128 @@ def test_excelwriter_fspath(self):
+
+
+@td.skip_if_no('odf')
+class TestODFReader(SharedItems):
```
Hey @detrout ! So we just went through a drastic overhaul of the test cases. You will want to definitely merge master in before continuing on this which should (hopefully) make adding these tests easier.

Rather than having everything in one large `test_excel` file you'll now see an excel subfolder where that used to be. You will want to do add this as `test_odf` as a separate module to start. We'll also have to figure out how to add this into the standard `test_readers` class but let's just take care of the former here and we can go from there

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd committed:
Added test_odf

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd committed:
Class naming consistency

WillAyd committed:
Whatsnew linting

WillAyd committed:
Added optional dependency load

WillAyd committed:
typo

WillAyd committed:
Updated inheritance to use excel reader interface

WillAyd committed:
Added ods test files

WillAyd committed:
Updated tests

WillAyd commented:
@detrout we've had some major changes to the Excel implementations and tests so I've merged master to keep these moving. I also changed the implementation you've provided to adhere strictly to the Excel reader interface defined in `_base.py`

There's some more cleanup that needs to be done which will be easier after #27114 gets merged (tests will fail for now) but give this a look and let me know if you have any questions on the direction

WillAyd committed:
convert_float handling

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd committed:
Fixed missing value handling

WillAyd committed:
Fixed error handling

WillAyd committed:
Fixed bool handling

WillAyd committed:
Skip missing file on master

WillAyd committed:
datetime compat

WillAyd committed:
fixed row repeat

WillAyd committed:
multiindex handling

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd committed:
Handled horizontally merged cells

WillAyd committed:
Converted to pytest idiom

WillAyd committed:
Test idiom cleanup

WillAyd committed:
Removed duplicative test files

WillAyd committed:
Raised NotImplemented for vertical merging

WillAyd committed:
Table attribute access simplification

WillAyd committed:
Typing and func cleanups

WillAyd committed:
lint and isort

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd committed:
typing fixup

WillAyd committed:
Skip ods files for xlrd

WillAyd commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,192 @@
+        max_row_len = 0
+        row_spans = {}  # type: Dict[int, int]
+
+        for i, sheet_row in enumerate(sheet_rows):
```
@detrout had to modify this method a bit for compat with other readers and to handle vertical cell merging. My extensions to your code may not be ideal so please let me know of feedback

WillAyd replied:
Also a lot of the complication here is due to the sparse nature of the source file. @TomAugspurger may have general insights on how to correctly parse this type of data

WillAyd commented on pandas/tests/io/excel/test_odf.py:
```diff
@@ -0,0 +1,76 @@
+    sheet = pd.read_excel("lowerdiagonal.ods", 'Sheet1',
+                          index_col=None, header=None)
+
+    assert sheet.shape == (4, 4)
```
Might change this and subsequent tests below to use `tm.assert_frame_equal` in another iteration or PR

WillAyd commented on pandas/tests/io/excel/test_readers.py:
```diff
@@ -439,6 +445,9 @@ def test_bad_engine_raises(self, read_ext):
 
     @tm.network
     def test_read_from_http_url(self, read_ext):
+        if read_ext == '.ods':  # TODO: remove once on master
```
This test only works when the file is available on master, so have to merge first and then can try again

alimcmaster1 replied:
Seems like this code is still hanging around ill aim to address it when I tackle: https://github.com/pandas-dev/pandas/issues/29439

WillAyd reviewed (commented):
@jreback mostly green with good coverage here. Travis failures are a result of `defusedxml` and `xlrd` being installed alongside one another and not sure there are great options to change. Should just catch warning in tests?

WillAyd committed:
Removed one-off tests

WillAyd committed:
Handled defusedxml warnings

WillAyd committed:
Updated assert_warnings funcs to allow DeprecationWarnings

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd commented:
@jreback all green here if you want to take a look for 0.25

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -164,6 +164,7 @@ Other enhancements
 - Added new option ``plotting.backend`` to be able to select a plotting backend different than the existing ``matplotlib`` one. Use ``pandas.set_option('plotting.backend', '<backend-module>')`` where ``<backend-module`` is a library implementing the pandas plotting API (:issue:`14130`)
 - :class:`pandas.offsets.BusinessHour` supports multiple opening hours intervals (:issue:`15481`)
 - :func:`read_excel` can now use ``openpyxl`` to read Excel files via the ``engine='openpyxl'`` argument. This will become the default in a future release (:issue:`11499`)
+- :func:`pandas.io.excel.read_excel` supports reading OpenDocument tables. Specify engine='odf' to enable. (:issue:`9070`)
```
can you put ``enable='odf'`` in double back ticks

jreback commented on pandas/io/excel/_odfreader.py:
```diff
@@ -0,0 +1,192 @@
+            sheet_cells = sheet_row.getElementsByType(TableCell)
+            empty_cells = 0
+            table_row = []  # type: List[Scalar]
+
```
may want to break this up with some helper methods (e.g. leave the nested loops here ) and call into a function for a cell

WillAyd replied:
The tricky thing with moving the cell-level code out is going to be the `row_span` (which signals if a particular cell is merged vertically). This gets passed to a dict whose scope is outside of the loops so that the next row iteration can know if it needs to account for a vertical merge while de-sparsifying

So if the cell-level parsing gets moved to a separate function I think we'd want to return the row_span from that or pass all of the mutable structures into that function. Not sure if that helps readability

^ not objecting just a point to consider

WillAyd replied:
Ignore this comment - found a better way

WillAyd replied:
Refactor up so hope this is less of an issue. The inner loops both modify containers declared outside of their respective loops so I *think* moving might make things less readable but lmk

jreback reviewed (changes requested):
are the config_init.py options updated? can you add a note in io.rst about this & update install.rst?

WillAyd committed:
Updated to config_init.py

WillAyd committed:
Updated whatsnew

WillAyd committed:
Updated io.rst

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

WillAyd reviewed (commented):
Documentation updates done. Dialog on how to best represent parsing

WillAyd committed:
Refactored to simplify

WillAyd committed:
Removed unnecessary test

WillAyd committed:
lint fixup

detrout commented:
Hi, ... sorry I was busy...

I finally had a chance to start reading what @WillAyd had done. (Thank you! that was a lot of updating)

I've only read up to 8a9a66c39a0adcc75540a2f6c492929cd2ec303c and want to read through the rest of the changes.

WillAyd committed:
mypy error

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into libreoffice-support

jreback commented on doc/source/user_guide/io.rst:
```diff
@@ -3217,7 +3219,20 @@ The look and feel of Excel worksheets created from pandas can be modified using
 * ``freeze_panes`` : A tuple of two integers representing the bottommost row and rightmost column to freeze. Each of these parameters is one-based, so (1, 1) will freeze the first row and first column (default ``None``).
 
+.. _io.ods:
 
```
can you add a versionchanged here

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -164,6 +164,7 @@ Other enhancements
 - Added new option ``plotting.backend`` to be able to select a plotting backend different than the existing ``matplotlib`` one. Use ``pandas.set_option('plotting.backend', '<backend-module>')`` where ``<backend-module`` is a library implementing the pandas plotting API (:issue:`14130`)
 - :class:`pandas.offsets.BusinessHour` supports multiple opening hours intervals (:issue:`15481`)
 - :func:`read_excel` can now use ``openpyxl`` to read Excel files via the ``engine='openpyxl'`` argument. This will become the default in a future release (:issue:`11499`)
+- :func:`pandas.io.excel.read_excel` supports reading OpenDocument tables. Specify ``engine='odf'`` to enable. (:issue:`9070`)
```
can you add a link to the new docs

jreback commented on doc/source/user_guide/io.rst:
```diff
@@ -3217,7 +3219,20 @@ The look and feel of Excel worksheets created from pandas can be modified using
+using the ``odfpy`` module. The semantics and features for reading
+OpenDocument spreadsheets match what can be done for `Excel files`_ using
+``engine='odf'``.
+
```
can you show an example here (code is ok)

jreback commented on pandas/tests/io/excel/test_readers.py:
```diff
@@ -62,14 +79,16 @@ def test_usecols_int(self, read_ext, df_ref):
         with tm.assert_produces_warning(FutureWarning,
-                                        check_stacklevel=False):
+                                        check_stacklevel=False,
+                                        raise_on_extra_warnings=False):
```
why does this have an extra warning?

WillAyd replied:
The installation of defusedxml (required by odfpy) unfortunately triggers `DeprecationWarning` and/or `PendingDeprecationWarning` during the XML parsing of ODS and Excel files. This is accounted for at a higher level during the parametrizataion of `read_engine` but needs to be allowed here as well or else this throws an error

jreback reviewed (changes requested):
lgtm

WillAyd committed:
Doc updates

jreback commented:
thanks @detrout and @WillAyd for the review and patches; very nice feature!

detrout commented:
Thank you @WillAyd and @jreback I'm glad this made it in!
